Users/jessezhu/deltaquerydirectoryobjects (#19994)
microsoftgraph/microsoft-graph-docs: api-reference/beta/api/directoryobject-delta.md

| Delegated (work or school account)     | Must have the read permission of each type specified in the $filter parameter. |
| Delegated (personal Microsoft account) | Not supported.                                                                                         |
| Application                            | Must have the read permission of each type specified in the $filter parameter. |

> **Note:** The `$filter` parameter is required, along with the read all permissions of each type specified. For example, for `https://graph.microsoft.com/beta/directoryObjects/delta/?$filter=isof('microsoft.graph.user') or isof('microsoft.graph.application')'`, both User.Read.All and Application.Read.All are required. 

## HTTP request

To begin tracking changes, you make a request including the delta function on the directoryObjects resource, along with the required $filter parameter.

<!-- { "blockType": "ignored" } -->

```http
GET /directoryObjects/delta?$filter=isof('{resource type}')
```

or 

```http
GET /directoryObjects/delta?$filter=id eq '{id}'
```

> ## Query parameters
> 
> Tracking changes in directoryObjects incurs a round of one or more **delta** function calls. If you use any query parameter (other than `$deltatoken` and `$skiptoken`), you must specify it in the initial **delta** request. Microsoft Graph automatically encodes any specified parameters into the token portion of the `@odata.nextLink` or `@odata.deltaLink` URL provided in the response.
> 
> You only need to specify any desired query parameters once upfront.
> 
> In subsequent requests, copy and apply the `@odata.nextLink` or `@odata.deltaLink` URL from the previous response, as that URL already includes the encoded, desired parameters.
> 
> | Query parameter | Type   | Description                                                                                                                                                                                                                                                                                                                                                   |
> | :-------------- | :----- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
> | $deltatoken     | string | A [state token](/graph/delta-query-overview) returned in the `@odata.deltaLink` URL of the previous **delta** function call for the same group collection, indicating the completion of that round of change tracking. Save and apply the entire `@odata.deltaLink` URL including this token in the first request of the next round of change tracking for that collection. |
> | $skiptoken      | string | A [state token](/graph/delta-query-overview) returned in the `@odata.nextLink` URL of the previous **delta** function call, indicating there are further changes to be tracked in the same group collection.                                                                                                                                                         |

Reviewer: I don't see this info being valuable here. It's valuable in the response body section. I recommend removing it from here.
```suggestion
```

### OData query parameters

This method supports optional OData query parameters to help customize the response.

> - You can use a `$select` query parameter as in any GET request to specify only the properties your need for best performance. The _id_ property is always returned. This must be accompanied with the required `$filter` parameter.
> - There is limited support for `$filter`:
>   - For tracking changes on a specific object: `$filter=id+eq+{value}`. You can filter multiple objects. For example, `https://graph.microsoft.com/beta/groups/delta/?$filter= id eq '477e9fc6-5de7-4406-bb2a-7e5c83c9ffff' or id eq '004d6a07-fe70-4b92-add5-e6e37b8affff'`. There is a limit of 50 filtered objects.
>   - For tracking changes on a specific resource: `$filter=isof({resource})`. For example, `https://graph.microsoft.com/beta/groups/delta/?$filter=isof('microsoft.graph.user') or isof('microsoft.graph.application')'`.
>   - Note: these filter types can be combined, e.g. `$filter=isof({resource type}) or id eq '{id}`, that provides an **intersection** of objects specified by `{resource type}` and `{id}`. 

Reviewer: This API behaves differently compared to the other delta, like /users/delta. Please check the technical accuracy and relevance of these statements. For example, URLs shouldn't be /groups/delta, but /directoryObjects/delta, et al.
```suggestion
- You can use a `$select` query parameter as in any GET request to specify only the properties your need for best performance. The _id_ property is always returned. This must be accompanied with the required `$filter` parameter.
- There is limited support for `$filter`:
  - For tracking changes on a specific object: `$filter=id+eq+{value}`. You can filter multiple objects. For example, `https://graph.microsoft.com/beta/groups/delta/?$filter= id eq '477e9fc6-5de7-4406-bb2a-7e5c83c9ffff' or id eq '004d6a07-fe70-4b92-add5-e6e37b8affff'`. There is a limit of 50 filtered objects.
  - For tracking changes on a specific resource: `$filter=isof({resource})`. For example, `https://graph.microsoft.com/beta/groups/delta/?$filter=isof('microsoft.graph.user') or isof('microsoft.graph.application')'`.
  - Note: these filter types can be combined, e.g. `$filter=isof({resource type}) or id eq '{id}`, that provides an **intersection** of objects specified by `{resource type}` and `{id}`. 
```

#### Default: return the same properties as initial delta request

By default, requests using a `@odata.deltaLink` or `@odata.nextLink` return the same properties as selected in the initial delta query in the following ways:

- If the property has changed, the new value is included in the response. This includes properties being set to null value.
- If the property has not changed, the old value is included in the response.
- If the property has never been set before it will not be included in the response at all.

> **Note:** With this behavior, by looking at the response it is not possible to tell whether a property is changing or not. Also, the delta responses tend to be large because they contain all property values - as shown in the [second example](#request-3) below.

#### Alternative: return only the changed properties

Adding an optional request header - `prefer:return=minimal` - results in the following behavior:

- If the property has changed, the new value is included in the response. This includes properties being set to null value.
- If the property has not changed, the property is not included in the response at all. (Different from the default behavior.)

> > **Note:** The header can be added to a `@odata.deltaLink` request at any point in time in the delta cycle. The header only affects the set of properties included in the response and it does not affect how the delta query is executed. See the [third example](#request-4) below.

Reviewer: I guess this can remain here for now. I'll work with George do see if this information should alternatively be in the Delta overview instead.

> ### Example

Reviewer: ```suggestion
## Examples

```

> #### Request 1
> 
> The following is an example of the request using the `$filter=isof('{resource type}')` parameter on users and groups. There is no `$select` parameter, so a default set of properties is tracked and returned.

Reviewer: ```suggestion
### Example 1: Retreve changes for a collection of users and groups

#### Request

The following is an example of the request using the `$filter=isof('{resource type}')` parameter on users and groups. There is no `$select` parameter, so a default set of properties is tracked and returned.
```

<!-- {
  "blockType": "request",
  "name": "directoryObject_delta_filter_isof"
}-->

```msgraph-interactive
GET https://graph.microsoft.com/beta/directoryObjects/delta?filter=isof('microsoft.graph.user') or isof('microsoft.graph.group')
```

> ---

Reviewer: nit
```suggestion
```

> #### Response 1

Reviewer: ```suggestion
#### Response
```

> The following is an example of the response when using `@odata.deltaLink` obtained from the query initialization with `$filter=isof('{resource type}')`:
> 
> > **Note:** The response object shown here might be shortened for readability.
> >
> > Note the presence of the _members@delta_ property which includes the ids of member objects in the group.

Reviewer: ```suggestion
The following is an example of the response when using `@odata.deltaLink` obtained from the query initialization with `$filter=isof('{resource type}')`. Note the presence of the _members@delta_ property which includes the ids of member objects in the group.

> **Note:** The response object shown here might be shortened for readability.
```

  "value":[
    {
      "@odata.type": "#microsoft.graph.user",
      "businessPhones": [
          "+1 425 555 0109"
      ],
      "displayName": "Adele Vance",
      "givenName": "Adele",
      "jobTitle": "Retail Manager",
      "mail": "AdeleV@contoso.onmicrosoft.com",
      "mobilePhone": "+1 425 555 0109",
      "officeLocation": "18/2111",
      "preferredLanguage": "en-US",
      "surname": "Vance",
      "userPrincipalName": "AdeleV@contoso.onmicrosoft.com",
      "id": "87d349ed-44d7-43e1-9a83-5f2406dee5bd"
    },
    {
      "@odata.type": "#microsoft.graph.group",
      "createdDateTime":"2021-03-12T10:36:14Z",
      "description":"This is the default group for everyone in the network",
      "displayName":"All Company",
      "groupTypes": [
        "Unified"
      ],
      "mail": "allcompany@contoso.com",
      "id": "72052a9a-c466-4995-8210-95a1c1221995",
      "members@delta": [
        {
          "@odata.type": "#microsoft.graph.user",
          "id": "693acd06-2877-4339-8ade-b704261fe7a0"
        },
        {
          "@odata.type": "#microsoft.graph.user",
          "id": "49320844-be99-4164-8167-87ff5d047ace"
        }
      ]
    }
  ]
}
```

> #### Request 2

Reviewer: ```suggestion
### Example 2: Retrieve a collection of changes for a directory object

#### Request
```

> The next example shows the request using the `$filter=id eq '{id}'` parameter on a group id. There is no `$select` parameter, so a default set of properties is tracked and returned.

Reviewer: ```suggestion
The following example shows the request using the `$filter=id eq '{id}'` parameter. There is no `$select` parameter, so a default set of properties is tracked and returned.
```

<!-- {
  "blockType": "request",
  "name": "directoryObject_delta_filter_id"
}-->

```msgraph-interactive
GET https://graph.microsoft.com/beta/directoryObjects/delta?$filter=id eq '87d349ed-44d7-43e1-9a83-5f2406dee5bd'
```

> ---

Reviewer: ```suggestion
```

> #### Response 2

Reviewer: ```suggestion
#### Response
```

<!-- {
  "blockType": "response",
  "truncated": true,
  "@odata.type": "microsoft.graph.directoryObject",
  "isCollection": true
} -->

```http
HTTP/1.1 200 OK
Content-type: application/json

{
  "@odata.context":"https://graph.microsoft.com/beta/$metadata#directoryObjects",
  "@odata.nextLink":"https://graph.microsoft.com/beta/directoryObjects/delta?$skiptoken=pqwSUjGYvb3jQpbwVAwEL7yuI3dU1LecfkkfLPtnIjsXoYQp_dpA3cNJWc",
  "value": [
    {
      "@odata.type": "#microsoft.graph.group",
      "createdDateTime":"2021-03-12T10:36:14Z",
      "description":"This is the default group for everyone in the network",
      "displayName":"All Company",
      "groupTypes": [
        "Unified"
      ],
      "mail": "allcompany@contoso.com",
      "id": "72052a9a-c466-4995-8210-95a1c1221995",
      "members@delta": [
        {
          "@odata.type": "#microsoft.graph.user",
          "id": "693acd06-2877-4339-8ade-b704261fe7a0"
        },
        {
          "@odata.type": "#microsoft.graph.user",
          "id": "49320844-be99-4164-8167-87ff5d047ace"
        }
      ]
    }
  ]
}
```

> #### Request 3
> 
> The next example shows the initial request selecting one property each from a user and group for change tracking, with default response behavior:

Reviewer: ```suggestion
### Example 3: Retrieve changes to specific properties for a collection of users and groups

#### Request

The following example shows the initial request selecting one property each from a user and group for change tracking, with default response behavior:
```

<!-- {
  "blockType": "request",
  "name": "directoryObject_delta_with_select"
}-->

```msgraph-interactive
GET https://graph.microsoft.com/beta/directoryObjects/delta?$filter=isof('microsoft.graph.user') or isof('microsoft.graph.group')&$select=microsoft.graph.user/surname,microsoft.graph.group/displayName
```

> ---

Reviewer: ```suggestion
```

#### Response 3

The following is an example of the response when using `@odata.deltaLink` obtained from the query initialization. Note that both properties are included in the response and it is not known which ones have changed since the `@odata.deltaLink` was obtained.

<!-- {
  "blockType": "response",
  "truncated": true,
  "@odata.type": "microsoft.graph.directoryObject",
  "isCollection": true
} -->

```http
HTTP/1.1 200 OK
Content-type: application/json

{
  "@odata.context":"https://graph.microsoft.com/beta/$metadata#directoryObjects",
  "@odata.nextLink":"https://graph.microsoft.com/beta/directoryObjects/delta?$skiptoken=pqwSUjGYvb3jQpbwVAwEL7yuI3dU1LecfkkfLPtnIjsXoYQp_dpA3cNJWc",
  "value": [
    {
      "@odata.type": "#microsoft.graph.user",
      "surname": "Vance",
      "id": "87d349ed-44d7-43e1-9a83-5f2406dee5bd"
    },
    {
      "@odata.type": "#microsoft.graph.group",
      "displayName":"All Company",
      "id": "72052a9a-c466-4995-8210-95a1c1221995",
    }
  ]
}
```

> #### Request 4
> 
> The next example shows the initial request selecting one property each from a user and group for change tracking, with alternative minimal response behavior:

Reviewer: ```suggestion
### Example 4: Retrieve specific properties only if they changed for a collection of users and groups

#### Request

The following example shows the initial request selecting one property each from a user and group for change tracking, with alternative minimal response behavior:
```

<!-- {
  "blockType": "request",
  "name": "directoryObject_delta_minimal"
}-->

```msgraph-interactive
GET https://graph.microsoft.com/beta/directoryObjects/delta?$filter=isof('microsoft.graph.user') or isof('microsoft.graph.group')&$select=microsoft.graph.user/surname,microsoft.graph.group/displayName
Prefer: return=minimal
```

> ---

Reviewer: ```suggestion
```

> #### Response 4

Reviewer: ```suggestion
#### Response
```

| [administrativeUnit](../resources/administrativeunit.md)     | AdministrativeUnit.Read.All | AdministrativeUnit.Read.All | Not supported |
| [appRoleAssignment](../resources/approleassignment.md)     | AppRoleAssignment.Read.All | AppRoleAssignment.Read.All | Not supported |
| [device](../resources/device.md)     | Device.Read.All | Device.Read.All | Not supported |
| [directoryRole](../resources/directoryrole.md)     | DirectoryRole.Read.All | DirectoryRole.Read.All | Not supported |
| [group](../resources/group.md)         | Group.Read.All | Group.Read.All | Not supported |
| [orgContact](../resources/orgcontact.md)     | OrgContact.Read.All | OrgContact.Read.All | Not supported |
| [servicePrincipal](../resources/serviceprincipal.md)     | ServicePrincipal.Read.All | ServicePrincipal.Read.All | Not supported |
| [user](../resources/user.md)           | User.Read.All | User.Read.All | Not supported |

## HTTP request

Track changes for a collection of a directory object type.

<!-- { "blockType": "ignored" } -->
```http
GET /directoryObjects/delta?$filter=isof('microsoft.graph.application')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.administrativeUnit')

> GET /directoryObjects/delta?$filter=isof('microsoft.graph.appRoleAssignment')

Reviewer: @jzhu2017 - Please replace the appRoleAssignment syntax with the full syntax as per your previous comment

> **Note:** To complete a delta query on the `appRoleAssignment` resource, both the `appRoleAssignment` and `user` resources must be specified in the $filter parameter.

| [orgContact](../resources/orgcontact.md)     | OrgContact.Read.All | OrgContact.Read.All | Not supported |
| [servicePrincipal](../resources/serviceprincipal.md)     | ServicePrincipal.Read.All | ServicePrincipal.Read.All | Not supported |
| [user](../resources/user.md)           | User.Read.All | User.Read.All | Not supported |

## HTTP request

Track changes for a collection of a directory object type.

<!-- { "blockType": "ignored" } -->
```http
GET /directoryObjects/delta?$filter=isof('microsoft.graph.application')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.administrativeUnit')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.appRoleAssignment') or isof('microsoft.graph.user')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.device')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.directoryRole')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.group')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.orgContact')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.servicePrincipal')
GET /directoryObjects/delta?$filter=isof('microsoft.graph.user')
```

> > **Note:** To complete a delta query on the `appRoleAssignment` resource, at minimum both the `appRoleAssignment` and `user` resources must be specified in the $filter parameter, with the option of also including groups or service principals if desired. Note that all resources specified will be returned.

Reviewer: Given you've already updated the syntax in Line 42 above, you can delete this line
```suggestion
```
